### PR TITLE
introducing cue for product configuration definition and validation

### PR DESCRIPTION
**Pull request by souleb**
## Description

introducing cue for product configuration definition and validation

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related Issue # (issue)
- Closes # (issue)
- Fixes # (issue)
> Remove if not applicable

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🙅 no documentation needed

## Checklist:

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api/v1alpha1/productdescription_types.go [hunk 1]
```diff
@@ -35,23 +35,20 @@ type ProductDescriptionPipeline struct {
...
+	//+required
+	Schema ResourceReference `json:"schema,omitempty"`
 }
 
 // DescriptionConfiguration contains details one parsing configuration items in a project description.
 type DescriptionConfiguration struct {
 	// +required
 	Rules ResourceReference `json:"rules"`
-	// +required
-	Readme ResourceReference `json:"readme"`
```
Comment (Skarlso) [inline api/v1alpha1/productdescription_types.go:54]: Don't we still want the README to provide some human readable information on how to use configs?

File: controllers/validation_controller.go [hunk 2]
```diff
@@ -1,323 +0,0 @@
...
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&mpasv1alpha1.Validation{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
-		Owns(&sourcebeta2.GitRepository{}).
-		Complete(r)
-}
-
-//+kubebuilder:rbac:groups=mpas.ocm.software,resources=validations,verbs=get;list;watch;create;update;patch;delete
-//+kubebuilder:rbac:groups=mpas.ocm.software,resources=validations/status,verbs=get;update;patch
-//+kubebuilder:rbac:groups=mpas.ocm.software,resources=validations/finalizers,verbs=update
-//+kubebuilder:rbac:groups=mpas.ocm.software,resources=repositories,verbs=get;list;watch;create;update;patch;delete
```
Comment (Skarlso) [inline controllers/validation_controller.go:74]: I think we still need the repositories permissions.. 🤔

File: controllers/productdeploymentgenerator_controller.go [hunk 3]
```diff
@@ -453,89 +410,119 @@ func (r *ProductDeploymentGeneratorReconciler) createProductDeployment(
 
 		readme = append(readme, []byte(fmt.Sprintf("\n# Configuration instructions for %s\n\n", p.Name))...)
 
-		parsed, err := r.increaseHeaders(instructions)
+		parsed, err := r.increaseHeaders([]byte(file.Comments()))
```
Comment (Skarlso) [inline controllers/productdeploymentgenerator_controller.go:413]: Could you please explain with a comment why this is needed? :) My fucking brain would forget about it eventually.

File: controllers/productdeploymentgenerator_controller.go [hunk 4]
```diff
@@ -453,89 +410,119 @@ func (r *ProductDeploymentGeneratorReconciler) createProductDeployment(
...
+		return nil, fmt.Errorf("failed to format config: %w", err)
 	}
 
-	if err := os.WriteFile(filepath.Join(dir, "values.yaml"), valuesYaml, 0o777); err != nil {
+	if err := os.WriteFile(filepath.Join(dir, "config.cue"), configBytes, 0644); err != nil {
 		return nil, fmt.Errorf("failed to write configuration values file: %w", err)
 	}
 
-	if err := os.WriteFile(filepath.Join(dir, "README.md"), readme, 0o777); err != nil {
+	if err := os.WriteFile(filepath.Join(dir, "README.md"), readme, 0644); err != nil {
```
Comment (Skarlso) [inline controllers/productdeploymentgenerator_controller.go:471]:
```suggestion
	if err := os.WriteFile(filepath.Join(dir, "README.md"), readme, 0o644); err != nil {
```

File: controllers/productdeploymentgenerator_controller.go [hunk 5]
```diff
@@ -453,89 +410,119 @@ func (r *ProductDeploymentGeneratorReconciler) createProductDeployment(
...
+	}
+
 	productDeployment.Spec = spec
 
 	logger.Info("successfully generated product deployment", "productDeployment", klog.KObj(productDeployment))
 
 	return productDeployment, nil
 }
 
+func (r *ProductDeploymentGeneratorReconciler) generateConfigMap(ctx context.Context, name, namespace, config string) error {
```
Comment (Skarlso) [inline controllers/productdeploymentgenerator_controller.go:492]: Do we want to set the owner so it gets removed if no longer needed?

File: api/v1alpha1/productdescription_types.go (hunk 1 above)
Comment (souleb) [reply api/v1alpha1/productdescription_types.go:54]: We put everything in the comment of the schema.cue. The readme.md in the project repo will be generated from that.

File: controllers/productdeploymentgenerator_controller.go (hunk 5 above)
Comment (souleb) [reply controllers/productdeploymentgenerator_controller.go:492]: we do that during creation.

Comment (Skarlso) [reply controllers/productdeploymentgenerator_controller.go:497]: Hm...? Where?

**Review by phoban01 — COMMENTED** [review]
With a schema such as the following:

```
// This a config file for the product deployment pipeline.
// It is named config.cue.

#SchemaVersion: "v1.0.0"

podinfo: {
  // this field has a default value of 2
  replicas: *2 | int

  // this is a required field with of type string with a constraint
  redis_url!: string & =~ "^https://"
      
  // this is an optional field
  labels?: {[string]:string}

  // this is a generated field that will not be exposed to in the config.cue file
  // part of the final configuration values
  max_replicas: replicas * 5 @private(true)
}
```

The required and optional fields indicated by (`!` and `?`) are omitted. 

The header comment in the file is also not present.

Actual output is:

```
podinfo: {
	// this field has a default value of 2
	replicas: 2
}
```

expected output:

```
// This a config file for the product deployment pipeline.
// It is named config.cue.

podinfo: {
  // this field has a default value of 2
  replicas: 2

  // this is a required field with of type string with a constraint
  redis_url: string

  // this is an optional field
  labels: {}
}
```

File: controllers/productdeploymentgenerator_controller.go (hunk 5 above)
Comment (souleb) [reply controllers/productdeploymentgenerator_controller.go:497]: oh my bad. Will add it.

**Comment by souleb** [comment]
I have created 2 follow up issues #71 and 72

**Review by morri-son — APPROVED** [review]

File: controllers/productdeploymentgenerator_controller.go [hunk 6]
```diff
@@ -450,102 +406,158 @@ func (r *ProductDeploymentGeneratorReconciler) createProductDeployment(
...
 
 	return productDeployment, nil
 }
 
+func (r *ProductDeploymentGeneratorReconciler) generateConfigMap(ctx context.Context, obj *v1alpha1.ProductDeploymentGenerator, values string) error {
+
+	configMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			// TODO@souleb: The config name should be  a hash of the values
+			// and should gc with a max of configmap and by timestamp
```
Comment (Skarlso) [inline controllers/productdeploymentgenerator_controller.go:524]: max of configmap of what? 🤔

File: controllers/productdeploymentgenerator_controller.go [hunk 7]
```diff
@@ -450,102 +406,158 @@ func (r *ProductDeploymentGeneratorReconciler) createProductDeployment(
...
+			Name:      obj.Name + "-values",
+			Namespace: obj.Namespace,
+		},
+	}
+
+	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, configMap, func() error {
```
Comment (Skarlso) [inline controllers/productdeploymentgenerator_controller.go:530]: If we create a new configmap on changes, do we really require an update flow?

**Review by phoban01 — APPROVED** [review]
It looks good to me, thanks for the huge effort @souleb.

I think we are not preserving the comments in the generated config.cue but that is something we can comeback to in a follow up.

**Comment by souleb** [comment]
closed in favour of https://github.com/open-component-model/mpas-product-controller/pull/89

**Pull request closed without merging**